server/automute.go

```diff
@@ -0,0 +1,182 @@
+	PreferenceNameAutomuteEnabled = "automute_enabled"
+)
+
+// enableAutomute mutes all of the user's linked channels and DMs and sets a preference to track that automute is
+// enabled. It assumes that the caller has checked that the user is both connected and has their primary platform
+// set to MS Teams.
+func (p *Plugin) enableAutomute(userID string) (bool, error) {
+	return p.setAutomuteEnabledForUser(userID, true)
+}
+
+// disableAutomute mutes all of the user's linked channels and DMs and sets a preference to track that automute is
+// disabled. It assumes that the user is either not connected to MS Teams or has their primary platform set to MM.
+func (p *Plugin) disableAutomute(userID string) (bool, error) {
+	return p.setAutomuteEnabledForUser(userID, false)
+}
+
+func (p *Plugin) setAutomuteEnabledForUser(userID string, automuteEnabled bool) (bool, error) {
+	if channelsAutomuted := p.getAutomuteIsEnabledForUser(userID); channelsAutomuted == automuteEnabled {
+		// We've already automuted all the users' channels, so there's nothing to do
+		return false, nil
+	}
+
+	var membersToMute []*model.ChannelMemberIdentifier
+
+	channels, appErr := p.API.GetChannelsForTeamForUser("", userID, true)
+	if appErr != nil {
+		return false, errors.Wrap(appErr, fmt.Sprintf("Unable to get channels for user %s to automute them", userID))
+	}
+
+	for _, channel := range channels {
+		if linked, err := p.canAutomuteChannel(channel); err != nil {
+			return false, err
+		} else if !linked {
+			continue
+		}
+
+		membersToMute = append(membersToMute, &model.ChannelMemberIdentifier{ChannelId: channel.Id, UserId: userID})
+	}
+
+	if err := p.setChannelMembersAutomuted(membersToMute, automuteEnabled); err != nil {
+		return false, errors.Wrap(err, fmt.Sprintf("Unable to automute all channels for linked user %s", userID))
+	}
+
+	if err := p.setAutomuteIsEnabledForUser(userID, automuteEnabled); err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (p *Plugin) setChannelMembersAutomuted(members []*model.ChannelMemberIdentifier, automuteEnabled bool) error {
+	notifyProps := getNotifyPropsForAutomute(automuteEnabled)
+
+	i := 0
+	perPage := 200
+	for i < len(members) {
+		start := i
+		end := i + perPage
+		if end > len(members) {
+			end = len(members)
+		}
+
+		page := members[start:end]
+
+		if appErr := p.API.PatchChannelMembersNotifications(page, notifyProps); appErr != nil {
+			return errors.Wrap(appErr, "Unable to patch notify props for automuting")
+		}
+
+		i += perPage
+	}
+
+	return nil
+}
+
+func getNotifyPropsForAutomute(automuteEnabled bool) map[string]string {
+	if automuteEnabled {
+		return map[string]string{
+			model.MarkUnreadNotifyProp: model.ChannelMarkUnreadMention,
+			NotifyPropAutomuted:        "true",
+		}
+	}
+
+	return map[string]string{
+		model.MarkUnreadNotifyProp: model.ChannelMarkUnreadAll,
+		NotifyPropAutomuted:        "false",
```

Comment:
It's unfortunate that this persists "indefinitely" -- is it possible to return to the original state without this prop?

Reply:
No, I don't think we can because of how this is a patch operation. Unless we were to add some special logic to the plugin API or the DB code, I don't think we could pass an undefined value here to make the DB remove this prop. If we find a way to do that though, I'd like to, but since we're reliant on server 9.5 at this point, we'd have to leave it as-is for now